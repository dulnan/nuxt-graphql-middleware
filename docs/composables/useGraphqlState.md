# useGraphqlState

This composable allows you to set fetch options for the useGraphqlQuery,
useAsyncGraphqlQuery and useGraphqlMutation composables. One common use case is
to pass custom request headers to the GraphQL middleware request.

::: warning

The state is only used for requests made from within a Nuxt app context (e.g.
pages, route middleware, etc.). Usually this is used to "pass" information from
the client/browser context to the middleware.

:::

::: code-group

```typescript [plugins/graphqlState.ts]
export default defineNuxtPlugin({
  name: 'my-state-plugin',
  // Makes sure that your plugin is executed after this plugin.
  // If it were to run before, then `state` would be null.
  dependsOn: ['nuxt-graphql-middleware-provide-state'],
  setup() {
    const state = useGraphqlState()

    // This is nullable because it's injected by a plugin.
    if (!state) {
      return
    }

    // Set fetch options for all GraphQL queries and mutations.
    state.fetchOptions = {
      headers: {
        CustomHeader: 'foobar',
      },
    }
  },
})
```

:::
