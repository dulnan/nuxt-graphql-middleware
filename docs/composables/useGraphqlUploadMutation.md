# useGraphqlUploadMutation

This composable is only available when setting `enableFileUploads` to `true` in
the module's configuration. It allows to upload files inside a mutation. The
implementation follows the
[GraphQL multipart request specification](https://github.com/jaydenseric/graphql-multipart-request-spec),
which is supported by a lot of GraphQL servers.

## Basic Usage

The composable handles the FormData part, so files can be directly provided
inside the mutation variables:

```typescript
async function upload(image: File) {
  const data = await useGraphqlUploadMutation('uploadImage', {
    image,
  })
}
```

Multiple files are also supported, both in the same field or in deeply nested
fields:

```typescript
const files = ref<File[]>([])

const data = await useGraphqlUploadMutation('uploadFiles', {
  elements: files.value.map((file) => {
    return {
      name: file.name,
      file: file,
    }
  }),
})
```

## Server Route

To support file uploads, when the feature is enabled, an additional server route
is added:

```
/api/graphql_middleware/upload/[name_of_mutation]
```

This route expects a `multipart/form-data` request. The
`useGraphqlUploadMutation` composable makes sure the data is sent in the correct
format.

The server route does not perform any validations on the provided data, for
example there is no file size limitation. This should be handled separately, for
example on the web server or by implementing a custom Nitro middleware.

In addition, the route does not save any of the files. It only validates that a
valid, existing mutation is used. It also makes sure that it's not possible to
send arbitrary operations.
