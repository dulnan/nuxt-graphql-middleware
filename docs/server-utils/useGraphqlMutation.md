# useGraphqlMutation()

::: warning

While this util has the same name as the composable it's a completely separate
method. In particular, it **does not use** any state set using the
`useGraphqlState` composable or the `graphqlMiddleware.clientOptions.ts` file.

:::

This util is auto-imported and available in a server (nitro) context. It's
function signature is identical to the
[useGraphqlMutation composable](/composables/useGraphqlMutation) composable
available in a Nuxt app context.

## Example

```typescript
import { getQuery } from 'h3'

export default defineEventHandler(async (event) => {
  const id = getQuery(event).id
  const data = await useGraphqlMutation('trackVisit', {
    id,
  })
  return data.data.success
})
```

## Client Context

Since the client context returned in
[buildClientContext()](/configuration/client-options) is only available in a
Nuxt app context you can manually pass the context when making a mutation with
the server util:

```typescript
import { getQuery } from 'h3'

export default defineEventHandler(async (event) => {
  const id = getQuery(event).id
  const data = await useGraphqlMutation({
    name: 'trackVisit',
    variables: {
      id,
    },
    clientContext: {
      language: 'de',
    },
  })
  return data.data.success
})
```
