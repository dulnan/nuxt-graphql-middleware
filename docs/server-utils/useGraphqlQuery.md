# useGraphqlQuery()

::: warning

While this util has the same name as the composable it's a completely separate
method. In particular, it **does not use** any state set using the
`useGraphqlState` composable or the `graphqlMiddleware.clientOptions.ts` file.

:::

This util is auto-imported and available in a server (nitro) context. It's
function signature is identical to the
[useGraphqlQuery composable](/composables/useGraphqlQuery) composable available
in a Nuxt app context.

## Example

```typescript
export default defineEventHandler(async () => {
  const data = await useGraphqlQuery('users')
  return data.data.users.map((v) => v.email)
})
```

## Client Context

Since the client context returned in
[buildClientContext()](/configuration/client-options) is only available in a
Nuxt app context you can manually pass the context when making a query with the
server util:

```typescript
export default defineEventHandler(async () => {
  const data = await useGraphqlQuery({
    name: 'users',
    clientContext: {
      language: 'de',
    },
  })
  return data.data.users.map((v) => v.email)
})
```
