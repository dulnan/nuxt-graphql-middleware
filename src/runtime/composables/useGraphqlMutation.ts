import {
  type GetMutationArgs,
  type MutationObjectArgs,
  type GetMutationResult,
  encodeContext,
} from './../helpers/composables'
import { performRequest } from './nuxtApp'
import { clientOptions } from '#nuxt-graphql-middleware/client-options'
import type { GraphqlResponse } from '#nuxt-graphql-middleware/response'
import type { Mutation } from '#nuxt-graphql-middleware/operation-types'

/**
 * Performs a GraphQL mutation.
 */
export function useGraphqlMutation<
  K extends keyof Mutation,
  R extends GetMutationResult<K>,
>(
  ...args: GetMutationArgs<K> | [MutationObjectArgs<K>]
): Promise<GraphqlResponse<R>> {
  const [name, body, fetchOptions = {}, overrideClientContext = {}] =
    typeof args[0] === 'string'
      ? [args[0], args[1], args[2]?.fetchOptions, args[2]?.clientContext]
      : [
          args[0].name,
          args[0].variables,
          args[0].fetchOptions,
          args[0].clientContext,
        ]

  const globalClientContext = clientOptions.buildClientContext
    ? clientOptions.buildClientContext()
    : {}

  return performRequest<R>('mutation', name, 'post', {
    ...fetchOptions,
    body,
    params: {
      ...(fetchOptions.params || {}),
      ...encodeContext({
        ...globalClientContext,
        ...overrideClientContext,
      }),
    },
  })
}
