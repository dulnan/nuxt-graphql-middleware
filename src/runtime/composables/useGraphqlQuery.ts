import {
  type GraphqlMiddlewareQueryName,
  type GetQueryArgs,
  type QueryObjectArgs,
  type GetQueryResult,
  encodeContext,
} from './shared'
import { buildRequestParams } from './../helpers'
import { performRequest } from './nuxtApp'
import { clientOptions } from '#graphql-middleware-client-options'
import type { GraphqlMiddlewareQuery } from '#nuxt-graphql-middleware/generated-types'
import type { GraphqlResponse } from '#graphql-middleware-server-options-build'

/**
 * Performs a GraphQL query.
 */
export function useGraphqlQuery<
  T extends GraphqlMiddlewareQueryName,
  R extends GetQueryResult<T, GraphqlMiddlewareQuery>,
>(
  ...args:
    | GetQueryArgs<T, GraphqlMiddlewareQuery>
    | [QueryObjectArgs<T, GraphqlMiddlewareQuery>]
): Promise<GraphqlResponse<R>> {
  const [
    name,
    variables,
    fetchOptions = {},
    graphqlCaching = {},
    overrideClientContext = {},
  ] =
    typeof args[0] === 'string'
      ? [
          args[0],
          args[1],
          args[2]?.fetchOptions,
          args[2]?.graphqlCaching,
          args[2]?.clientContext,
        ]
      : [
          args[0].name,
          args[0].variables,
          args[0].fetchOptions,
          args[0].graphqlCaching,
          args[0].clientContext,
        ]

  const globalClientContext = clientOptions.buildClientContext
    ? encodeContext(clientOptions.buildClientContext())
    : {}

  const clientContext = {
    ...globalClientContext,
    ...overrideClientContext,
  }

  return performRequest<R>(
    'query',
    name,
    'get',
    {
      params: {
        ...buildRequestParams(variables),
        ...clientContext,
      },
      ...fetchOptions,
    },
    graphqlCaching,
  )
}
