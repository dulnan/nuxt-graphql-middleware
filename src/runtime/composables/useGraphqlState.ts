import { useNuxtApp } from '#imports'
import type { NuxtApp } from '#app'
import type { GraphqlMiddlewareState } from './../types'

export const useGraphqlState = function (
  providedApp?: NuxtApp,
): GraphqlMiddlewareState | null {
  try {
    const app = providedApp || useNuxtApp()
    if (app.$graphqlState) {
      return app.$graphqlState
    }
  } catch {
    // Noop.
  }
  return null
}
