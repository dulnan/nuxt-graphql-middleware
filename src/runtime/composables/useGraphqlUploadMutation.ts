import {
  type GraphqlMiddlewareMutationName,
  type GetMutationArgs,
  type MutationObjectArgs,
  type GetMutationResult,
  getEndpoint,
  encodeContext,
} from './shared'
import type { GraphqlMiddlewareMutation } from '#nuxt-graphql-middleware/generated-types'
import { clientOptions } from '#graphql-middleware-client-options'
import { useGraphqlState } from '#imports'
import type { GraphqlResponse } from '#graphql-middleware-server-options-build'

/**
 * Builds the form data.
 */
function createFormData(variables: Record<string, any>): FormData {
  const formData = new FormData()
  formData.append('operations', '{}')
  const map: Record<string, string[]> = {}
  let fileIndex = 0

  // Iterate over the variables and collect the files.
  const traverseAndBuildMap = (obj: any, path: string): any => {
    if (obj instanceof File) {
      const fileKey = `${fileIndex++}`
      map[fileKey] = [path]
      formData.append(fileKey, obj)
      return null // Replace File with null for JSON.stringify
    }

    if (Array.isArray(obj)) {
      return obj.map((item, index) =>
        traverseAndBuildMap(item, `${path}.${index}`),
      )
    }

    if (typeof obj === 'object' && obj !== null) {
      const newObj: Record<string, any> = {}
      for (const key in obj) {
        newObj[key] = traverseAndBuildMap(obj[key], `${path}.${key}`)
      }
      return newObj
    }

    return obj
  }

  const cleanedVariables = traverseAndBuildMap(variables, 'variables')

  formData.append('variables', JSON.stringify(cleanedVariables))
  formData.append('map', JSON.stringify(map))

  return formData
}

/**
 * Performs a GraphQL upload mutation.
 */
export function useGraphqlUploadMutation<
  T extends GraphqlMiddlewareMutationName,
  R extends GetMutationResult<T, GraphqlMiddlewareMutation>,
>(
  ...args:
    | GetMutationArgs<T, GraphqlMiddlewareMutation>
    | [MutationObjectArgs<T, GraphqlMiddlewareMutation>]
): Promise<GraphqlResponse<R>> {
  const [name, variables, fetchOptions = {}, overrideClientContext = {}] =
    typeof args[0] === 'string'
      ? [args[0], args[1], args[2]?.fetchOptions, args[2]?.clientContext]
      : [
          args[0].name,
          args[0].variables,
          args[0].fetchOptions,
          args[0].clientContext,
        ]

  if (!variables) {
    throw new Error(
      'Using "useGraphqlUploadMutation" without variables is not supported.',
    )
  }

  const state = useGraphqlState()

  const formData = createFormData(variables)

  const globalClientContext = clientOptions.buildClientContext
    ? encodeContext(clientOptions.buildClientContext())
    : {}

  const clientContext = {
    ...globalClientContext,
    ...overrideClientContext,
  }
  return $fetch<GraphqlResponse<R>>(getEndpoint('upload', name), {
    ...(state && state.fetchOptions ? (state.fetchOptions as any) : {}),
    ...(fetchOptions || {}),
    params: {
      ...clientContext,
    },
    method: 'POST',
    body: formData,
  }).then((v) => {
    return {
      ...v,
      data: v?.data,
      errors: v?.errors || [],
    }
  })
}
