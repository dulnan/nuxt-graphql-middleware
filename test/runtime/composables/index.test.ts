import { describe, expect, test, vi } from 'vitest'
import { useGraphqlState } from './../../../src/runtime/composables/useGraphqlState'
import { useGraphqlMutation } from './../../../src/runtime/composables/useGraphqlMutation'
import { useGraphqlQuery } from './../../../src/runtime/composables/useGraphqlQuery'
import { useGraphqlUploadMutation } from './../../../src/runtime/composables/useGraphqlUploadMutation'

const useNuxtApp = function () {
  return {
    $graphqlState: {
      fetchOptions: {},
    },
  }
}

vi.stubGlobal('useNuxtApp', useNuxtApp)

vi.mock('#nuxt-graphql-middleware/helpers', () => {
  return {
    getEndpoint(operation: string, operationName: string) {
      return `/nuxt-graphql-middleware/${operation}/${operationName}`
    },
  }
})
vi.mock('#imports', () => {
  return {
    useRuntimeConfig: () => {
      return {
        public: {
          'nuxt-graphql-middleware': {
            serverApiPrefix: '/nuxt-graphql-middleware',
          },
        },
        graphqlMiddleware: {
          graphqlEndpoint: 'http//localhost/graphql',
        },
      }
    },
    useNuxtApp: () => {
      return {
        $graphqlState: {
          fetchOptions: {},
        },
      }
    },
    useAppConfig: () => {
      return {
        graphqlMiddleware: {},
      }
    },
    useGraphqlState: () => {
      return {
        fetchOptions: {},
      }
    },
  }
})

vi.mock('#nuxt-graphql-middleware/client-options', () => {
  return {
    clientOptions: {},
  }
})

const fetchMock = (endpoint: string, options: any) => {
  return Promise.resolve({
    data: undefined,
    options,
    endpoint,
  })
}

vi.stubGlobal('$fetch', fetchMock)

describe('useGraphqlQuery', () => {
  test('Performs a query', async () => {
    expect(await useGraphqlQuery('foobar')).toMatchSnapshot()
  })

  test('Performs a query with string variables', async () => {
    expect(
      await useGraphqlQuery('foobar', { stringVar: 'foobar' }),
    ).toMatchSnapshot()
  })

  test('Performs a query with string variables', async () => {
    expect(
      await useGraphqlQuery('foobar', { stringVar: 'foobar' }),
    ).toMatchSnapshot()
  })

  test('Performs a query with non-string variables', async () => {
    expect(await useGraphqlQuery('foobar', { numeric: 123 })).toMatchSnapshot()
  })

  test('Throws an error for invalid query names.', async () => {
    const result = await useGraphqlQuery(123).catch((e) => e)
    expect(result).toMatchSnapshot()
  })

  test('takes options into account', async () => {
    expect(
      await useGraphqlQuery(
        'foobar',
        { stringVar: 'foobar' },
        {
          fetchOptions: {
            params: {
              customParam: 'yes',
            },
          },
          clientContext: {
            language: 'fr',
          },
        },
      ),
    ).toMatchInlineSnapshot(`
      {
        "data": undefined,
        "endpoint": "/nuxt-graphql-middleware/query/foobar",
        "errors": [],
        "options": {
          "method": "get",
          "params": {
            "__gqlc_language": "fr",
            "customParam": "yes",
            "stringVar": "foobar",
          },
        },
      }
    `)
  })
})

describe('useGraphqlMutation', () => {
  test('Performs a mutation', async () => {
    expect(await useGraphqlMutation('foobar')).toMatchSnapshot()
  })

  test('Performs a mutation with variables', async () => {
    expect(
      await useGraphqlMutation('foobar', { user: 'foobar' }),
    ).toMatchSnapshot()
  })

  test('Throws an error for invalid mutation names.', async () => {
    const result = await useGraphqlMutation(123).catch((e) => e)
    expect(result).toMatchSnapshot()
  })

  test('takes options into account', async () => {
    const result = await useGraphqlMutation(
      'foobar',
      { stringVar: 'foobar' },
      {
        fetchOptions: {
          params: {
            customParam: 'yes',
          },
        },
        clientContext: {
          language: 'fr',
        },
      },
    )
    expect(result.options.body).toMatchInlineSnapshot(`
      {
        "stringVar": "foobar",
      }
    `)
    expect(result.options.params).toMatchInlineSnapshot(`
      {
        "__gqlc_language": "fr",
        "customParam": "yes",
      }
    `)
  })
})

describe('useGraphqlUploadMutation', () => {
  test('takes options into account', async () => {
    const result = await useGraphqlUploadMutation(
      'foobar',
      { stringVar: 'foobar' },
      {
        fetchOptions: {
          params: {
            customParam: 'yes',
          },
        },
        clientContext: {
          language: 'fr',
        },
      },
    )
    expect(result).toMatchInlineSnapshot(`
      {
        "data": undefined,
        "endpoint": "/nuxt-graphql-middleware/upload/foobar",
        "errors": [],
        "options": {
          "body": FormData {
            Symbol(state): [
              {
                "name": "operations",
                "value": "{}",
              },
              {
                "name": "variables",
                "value": "{"stringVar":"foobar"}",
              },
              {
                "name": "map",
                "value": "{}",
              },
            ],
          },
          "method": "POST",
          "params": {
            "__gqlc_language": "fr",
            "customParam": "yes",
          },
        },
      }
    `)
  })
})

describe('useGraphqlState', () => {
  test('Returns the state', () => {
    expect(useGraphqlState()).toEqual({
      fetchOptions: {},
    })
  })
})
