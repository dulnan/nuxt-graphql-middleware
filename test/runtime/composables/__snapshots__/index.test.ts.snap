// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useGraphqlMutation > Performs a mutation 1`] = `
{
  "data": undefined,
  "endpoint": "/nuxt-graphql-middleware/mutation/foobar",
  "errors": [],
  "options": {
    "body": undefined,
    "method": "post",
    "params": {},
  },
}
`;

exports[`useGraphqlMutation > Performs a mutation with variables 1`] = `
{
  "data": undefined,
  "endpoint": "/nuxt-graphql-middleware/mutation/foobar",
  "errors": [],
  "options": {
    "body": {
      "user": "foobar",
    },
    "method": "post",
    "params": {},
  },
}
`;

exports[`useGraphqlMutation > Throws an error for invalid mutation names. 1`] = `
{
  "data": undefined,
  "endpoint": "/nuxt-graphql-middleware/mutation/undefined",
  "errors": [],
  "options": {
    "body": undefined,
    "method": "post",
    "params": {},
  },
}
`;

exports[`useGraphqlQuery > Performs a query 1`] = `
{
  "data": undefined,
  "endpoint": "/nuxt-graphql-middleware/query/foobar",
  "errors": [],
  "options": {
    "method": "get",
    "params": {},
  },
}
`;

exports[`useGraphqlQuery > Performs a query with non-string variables 1`] = `
{
  "data": undefined,
  "endpoint": "/nuxt-graphql-middleware/query/foobar",
  "errors": [],
  "options": {
    "method": "get",
    "params": {
      "__variables": "{"numeric":123}",
    },
  },
}
`;

exports[`useGraphqlQuery > Performs a query with string variables 1`] = `
{
  "data": undefined,
  "endpoint": "/nuxt-graphql-middleware/query/foobar",
  "errors": [],
  "options": {
    "method": "get",
    "params": {
      "stringVar": "foobar",
    },
  },
}
`;

exports[`useGraphqlQuery > Performs a query with string variables 2`] = `
{
  "data": undefined,
  "endpoint": "/nuxt-graphql-middleware/query/foobar",
  "errors": [],
  "options": {
    "method": "get",
    "params": {
      "stringVar": "foobar",
    },
  },
}
`;

exports[`useGraphqlQuery > Throws an error for invalid query names. 1`] = `
{
  "data": undefined,
  "endpoint": "/nuxt-graphql-middleware/query/undefined",
  "errors": [],
  "options": {
    "method": "get",
    "params": {},
  },
}
`;
